### DISCUSSION TITLE
ethernet gadget mode?

### DISCUSSION THREAD

**Discussion by tinker2much** [gitbls/sdm#263] [Ideas]
I have an old pi zero that had originally been set up as an ethernet gadget - I had soldered a USB-A plug to it, and I could plug it into a host, access it from the host, and also share the host's internet access with it. 

At some point this stopped working. I've noticed this post on the Raspberry Pi Forum by user phattmatt that touts a current working way to get this going, using first boot and network manager configuration: [phattmatt url](https://forums.raspberrypi.com/viewtopic.php?t=376578). And the words "first boot" and "network manager" sure made me think that this is something up your alley, as well as it being the kind of thing you and sdm handle regularly - fiddly but systematic things that you can solve once and spare the rest of us the bother ever after.

As to alternate methods of setting up ethernet gadget mode, see the guide mentioned by Thagrol within that thread: [Thagrol url](https://github.com/thagrol/Guides/blob/main/ethernetgadget.pdf)

**Comment by gitbls** [comment]
I've had gadget mode on my ToDo list for a while, but it's never floated to the top of the list. Will reconsider, especially if there's sufficient interest...

**Comment by tinker2much** [comment]
Willing to test...

**Comment by tinker2much** [comment]
The phatmatt method linked to above is meant to be applied to the boot partition of a just-burned card. Brief mods to config.txt and to cmdline.txt, then a block of code to be inserted into the standard firstrun.sh. That messes with a udev rule and creates two nmconnections. Boot and you're in business. Instead of using the standard firstrun.sh, I had a card I had burned with sdm and already booted, so I made the hand modifications and cut the code into a temporary script and ran it with sudo after the fact. He also gives instructions for how to share internet from the host Mac to the pi. It all seemed to work fine. 
Free bonus: the wireless connections created during sdm (of course, here in the case of a zero, only useful if I attach a wifi dongle) still work if I DON'T plug the thing into a USB slot. So if I just want the thing to hang out on my network and be available to the rest of the pi's, I can do that too. 
For extra credit:  would it be possible to bridge the mini-network created between the pi and the Mac, with the regular network all of my other pi's are on, so the gadget pi could see all the rest of them?

**Comment by gitbls** [comment]
Looks pretty straightforward to implement phatmatt's guide in sdm via a new plugin. Of course you'll be able to use the plugin either during customization or burning, which is quite handy. 

sdm won't do anything about the Windows or Mac side of things, although I'll see if there are any helper Powershell scripts that would make sense for Windows. Mac? Maybe someone can contribute a shortcut once this is all working.

Current plan is to complete the sdm release in the next several days (it's just about ready to go), and then build the gadget plugin.

It will need some testing besides my "lab", thanks @tinker2much for volunteering. Do you happen to have a Mac? Our library of Apple devices doesn't include a Mac.

**Comment by tinker2much** [comment]
Yes, I have a Mac and will be happy to test out a new gadget plugin over there.

**Comment by gitbls** [comment]
`gadgetmode` plugin is available with sdm V13.1. I've tested it on Windows 11 and Pi4 tether hosts using a Pi02W. Should work for any/all conforming hosts and clients.
